### Homework submission by:
* Name:
* Student number:

# Assignment week 5: Numpy Arrays / Matplotlib for Arrays

This weeks we will go over the basics of Numpy arrays. Numpy arrays are used all over scientific software written in
Python. Numpy provides a fast array datatype and functions that act on it, libraries like Scipy and Astropy then use
these arrays to provide more specialized routines. We will use Jupyter (formerly known as Ipython Notebook) for our
homework this week. You must complete the assignments in this notebook and hand in the notebook containing your 
answers and code (clearly marked with your name and student number). 

### Important notes regarding the homework:
You are not to use any `Scipy` functions unless specifically told so. In particular you are **not allowed** to use:
* `scipy.misc.imread`
* `matplotlib.pyplot.imread`
* `scipy.ndimage.imread`

Numpy speeds up code only if you use it appropriately, writing "manual" loops will slow things down.

Your code must run, check that it does before handing in.


### Grading
* Handing in: 1 point
* Style (Code, Plots, General): 3 points
* Assignments
    * Part 1: 0.5 points
    * Part 2: 0.5 points
    * Part 3: 1.0 points
    * Part 4: 1.5 points
    * Part 5: 1.0 points
    * Part 6: 1.5 points

The usual penalty for handing in late is applied this week as well.

### Part 1: Reading an image
*This assignment is worth 0.5 points.*

* Download the photo provided at 
* **Assignment:** Write a function `imread` that can read an image from disk and convert it to a Numpy array.
    * Use the `Pillow` library to read the image. If you do not have Pillow installed, install it first.
    * Numpy has a function that converts an image to an array.
    * Answer the question below!

In [4]:
# Note this cell only contains imports that are needed for this exercise. 
# If you get an ImportError when you run this cell, make sure to install the required Python packages.
%matplotlib inline
from PIL import Image  # (this uses Pillow on a recent installation!)
from matplotlib import pyplot
import numpy as np

In [9]:
# Add code that uses the Pillow `Image` class to load an image, and numpy code to turn it into an array:


def imread(filename):
    """Read an image, return a Numpy array."""
    im = Image.open(filename)
    return np.asarray(im, dtype=np.uint8)


ar = imread('spacex-intelsat-35e.jpg')  # ar will be reused below

print 'Shape', ar.shape[:2]
print 'Channels', ar.shape[2]

 Shape (640, 427)
Channels 3


* **Qestion**: How many pixels and how many channels does this image have? How can you find out using only Numpy?
  Add code **below and outside** of `imread` to print the image size and number of channels.
* **Answer**: ...

### Part 2: Creating a grayscale image
The `make_grayscale` function below creates a grayscale version of your color image.
* **Assignment**: Adapt the `make_grayscale` function to either use `numpy.dot` or `numpy.matmul` to perform the calculation. Answer the question below.

In [ ]:
# Rewrite `make_grayscale` such that it uses `numpy.matmul` or `numpy.dot`.

def make_grayscale(im_array):
    """Take a RGB color image (in array form) and return a grayscale (single channel) image."""
    return 0.2126 * im_array[:, :, 0] + 0.7152 * im_array[:, :, 1] + 0.0722 * im_array[:, :, 2]
    
pyplot.imshow(make_grayscale(ar))
pyplot.axis('Off')

* **Question**: The image shows up with color, but has only one channel when you check it.
  Explain why this is the case and how you can fix it (where fix means to show the image as a grayscale one).
  Demonstrate your fix by adapting the code above.
* **Answer**: ...

### Part 3: Color channels (more on color maps)
The function below demonstrates how you can extract the individual channels in an RGB image and how you can plot
them using Matplotlib.

In [ ]:
def show_channels(im_array):
    """Plot the individual channels in an RGB image."""
    titles = ['Red channel', 'Green channel', 'Blue channel']

    #  prepare the figure/axes to plot to:
    fig, axes = pyplot.subplots(nrows=1, ncols=4)
    w, h = fig.get_size_inches()
    fig.set_size_inches(4 * w, h)
    
    # separate the channels, plot them
    for i, t in enumerate(titles):
        channel = numpy.zeros_like(im_array)
        channel[:, :, i] = ar[:, :, i]

        axes[i].imshow(channel)
        axes[i].set_title(t)
        axes[i].axis('Off')
        
        print t, 'has array shape', channel.shape
    
    # plot the original image for comparison
    axes[3].imshow(im_array)
    axes[3].set_title('Original image')
    axes[3].axis('Off')

show_channels(ar)

* **Assignment**: The function above seperates the channels by setting the other channels to zero.
  You must change this function to create individual color channels that have a shape of (640, 427) --- for instance,
  the blue channel image should not contain red and green channels set to zero. Make sure that
  the individual channels can be plotted correctly (this involves creating your own color mappings in Matplotlib).
* **Hint**:
    * Use `matplotlib.colors.LinearSegmentedColormap` to create color maps for the individual channels. You will 
      need three of these color maps.
    * The plot at the end should not change if you created the correct color maps.

### Part 4: Color channel histograms
* **Assignment**: In this exercise you should create histograms for the individual red, green, blue channels. 
* **Hint**:
    * Use the `matplotlib.pyplot.hist` function to generate and plot the histograms.
    * The image below gives an idea of what the histograms should look like, note that it intentionally leaves
      out the axes labels --- **you will have to create appropriate labels**. Answer he question below.

<img src="files/rgb-hist-hint.png">

In [ ]:
def plot_rgb_histograms(im_array):
    pass  # add your plotting code here

plot_rgb_histograms(ar)

* **Question**: When you create a histogram, you can choose the number of bins. What is the appropriate number in this
  case. Study the arrays you are using to come to your answer. Implement your choice in the code above, make sure that
  the labeling on the x axis is on and shows this choice.
* **Answer**: ...

### Part 5: Generating grids in Numpy
Numpy has functions to generate grids of values that can then be passed to functions and evaluated in one go.
Note how in the example below you do not have to write a loop over all values of x in the array created by `numpy.linspace` . By avoiding loops in Python you will create faster running programs.

In [ ]:
# example code
x = numpy.linspace(0, 2 * numpy.pi, 100)
y = numpy.sin(x)

pyplot.plot(x, y)
pyplot.xlabel('$x$')
pyplot.ylabel('$\sin(x)$')

* **Assigment**: In this exercise you will use the `numpy.meshgrid` and `numpy.linspace` to generate a grid on which
  to evaluate a mathematical function (defined below) and then create a plot using `matplotlib.pyplot.imshow`.
    * Evaluate the function $f(x, y) = \sin(\sqrt{(x - \pi)^2 + (y - \pi)^2}) \sin(\sqrt{(x - 5\pi)^2 + (y - 5\pi)^2}).$
    * Use the domain: $x \in [0, 12\pi], y \in [0, 12\pi]$
    * Make sure that the axes are labeled correctly, so that they show the domain over which the function was evaluated.
* **Hints**:
    * Use `numpy.sin` and not `math.sin`.
    * This function looks as follows (this example is using an arbitrary color map and leaving out the axes labels
      intentionally - **you, however, should add axis labels**):
    
<img src="files/func-example.png">

In [ ]:
# Add your code here (generate a grid, evaluate the function and plot it).


### Part 6: Unsharp masking
A technique that is oftern used to sharpen images is so-called unsharp masking. In this assignment you will build your
own implementation of Unsharp masking. (*This part of the assignment is intentionally a bit harder than the rest 
of it!*)

Assignment:
* Read the Wikipedia page on unsharp masking: https://en.wikipedia.org/wiki/Unsharp_masking
* Implement the algorithm that is described on that page.
* Your submission should contain the code that implements this algorithm (put it below).
* Your submission should contain two grayscale images: the original one and for comparison the unsharp-masked one.
  These images should be included in the report - to see how that is done, check the two example images above.

Hints:
* Building a version of the algorithm that works for black and white images is enough. If your implementation
  receives a color image you are allowed to turn that into a grayscale image instead (and use that as input).
* Matplotlib will scale your images as it sees fit (both in size and in color mapping). To check whether your
  implementation works it is better to save an PNG image to disk (not JPEG, as that is a "lossy" format) than
  to create plots using `matplotlib.pyplot.imshow`.
* It is convenient to work in floating point numbers for the intermediate steps of the algorithm.
* Saving to disk needs `uint8` as datatype; rescale the values if needed (!) and change the datatype of the 
  image array.
* It is useful to make plots/figures of the intermediate results. Be careful because Matplotlib will rescale
  inputs automatically (use `vmin` and `vmax` keyword arguments appropriately to override this behavior).
  
Note on handing in:
* Make sure it is easy to run your code on a test image.

In [ ]:
# Implement the unsharp mask method of image sharpening here.